### PR TITLE
scan report file name depends from report type

### LINKED ISSUE #1942: Add support for output_name parameter
<table>
      <tr>
        <td>
          
        </td>
        <td>
          Original issue by @sguillope - Imported from <a href='https://github.com/fastlane/scan/issues/27'>fastlane/scan#27</a>
        </td>
      </tr>
    </table>

It would be nice if it was possible to specify the name of the output file (similar to `gym`'s `output_name` parameter).
We are running our tests on Bamboo but the [Junit Parser plugin](https://confluence.atlassian.com/bamboo/junit-parser-289277056.html) only seems to parse files with an `.xml` extension. We currently have to manually rename the output file `report.junit` to have the `.xml` extension.

### PR DESCRIPTION
**Pull request by nfalco79**
Addresses #1942

Define a different default based on report type (e.g. junit wants xml file) otherwise fallback on previous behaviour.

### CODE REVIEW

File: scan/lib/scan/report_collector.rb [hunk 1]
```diff
@@ -38,7 +38,12 @@ def generate_commands(path, types: nil, output_file_name: nil)
           next
         end
 
-        file_name = "report.#{type}"
+        case type
+        when "junit"
+          file_name = "TEST-report.xml"
```
Comment (KrauseFx) [inline scan/lib/scan/report_collector.rb]: That's the default name for junit files?

Comment (nfalco79) [inline scan/lib/scan/report_collector.rb]: Tipycally name for junit (Surefire Test Report) report is TEST-[report name].xml so I think the best choose is use report as "report name" part.

**Comment by LeffelMania** [comment]
+1 for this PR since CircleCI isn't picking up the junit test output because it doesn't have a `.xml` extension (I believe). I think the type-to-extension logic deserves its own method though:

```
...
    extension = report_extension(type)
    file_name = "report.#{extension}"
...

def report_extension(type)
    case type
    when "junit"
        return "xml"
    else
        return type
    end
end
```
(my Ruby is novice, but that's the idea)

**Comment by nfalco79** [comment]
The extension logic based on report type is not the best way.
Remember that report was generated by xcpretty so output could have very disparate usages.
Thing (**for example**) if this action does not have a whitelist for the report type, we could specify also the json-compilation-database report that generates a special db used from some other action (for coverage and so on) and they excepted exactly a compilation.json file. Using the method above you must hope that the next actions have the capability to specify the input parameter otherwise action will not find the file it are looking for.
Using the default name expected as output facilitates writing lane saving configuration of input parameters / output of following actions.

No one prohibits even have a parameter for the output file name, but at this point instead of doing joins  between name and extension as the default behavior I would keep the report type parameter to decide a default name and obviously overwritable by output_file parameter which in this case must contain both name and extension to have maximum freedom.

**Keep in mind that these kind of changes on a widely used action as is this one, will force to review all lane script of a lot of used that would update**

**Comment by mfurtak** [comment]
Hi @nfalco79 - thanks for the contribution! 😄

While I agree with you and the original ticket that the JUnit report name would be better as `TEST-something.xml`, I don't think we can go forward with the approach here.

Because the `report.junit` filename has been the default for so long, we can't change it because some large number of people may currently be depending on that file name during their build process.

I think it would be best to introduce a new `ConfigItem` option that more explicitly describes the desired "output type" to "output file name" relationship. Please let me know if you'd like to discuss how that might look and work 👍

**Pull request closed without merging**